plutus-core/plutus-core/src/PlutusCore/Value.hs

```diff
@@ -91,8 +104,48 @@ fromList =
 
 normalize :: NestedMap -> NestedMap
 normalize = Map.filter (not . Map.null) . Map.map (Map.filter (/= 0))
+{-# INLINEABLE normalize #-}
 
 instance Pretty Value where
   pretty = pretty . fmap (bimap toText (fmap (first toText))) . toList
    where
     toText = Text.decodeLatin1 . Base64.encode
+
+{-| \(O(\log \max(m, k))\), where \(m\) is the size of the outer map, and \(k\) is
+the size of the largest inner map.
+-}
+insertCoin :: ByteString -> ByteString -> Integer -> Value -> Value
+insertCoin currency token amt (Value outer sizes size) = case Map.lookup currency outer of
```

Comment:
You also need to check if `amt == 0`. In that case, the operation is equivalent to `deleteCoin currency token`.